### ISSUE TITLE
Prep work for CE:FT merger into CE

### ISSUE THREAD

**Issue by Azumgi** [AthenaSulisMinerva/CombatExtendedFastTrack#251]
10 hours ago NoImageAvailable did a PSA in CE discord:
> *PSA:* As some of you may already know, I've not been very active in CE development since the 1.6 release. Unfortunately between working on Ascent of Ashes and some general health issues I just haven't had the time or energy to work actively on CE. This has not stopped other contributors from working on and submitting features and patches to the GitHub repository and while I kept meaning to review them I just never got around to it. In the end I felt bad over the backlog, authors weren't very happy with their work stuck in limbo and users didn't get new content. Losers all around.
> For this reason I've decided to officially *step down as maintainer of CE* so that someone more active may take over and give the mod the attention it requires. If you have the aptitude and are willing to take this responsibility please PM me for further details. Ideally I'd like to set the new maintainer up with full access to the current repository, rather than a fork. This way, all old download links will remain valid and we won't have users accidentally downloading outdated versions. Ideally users wouldn't notice anything of the switch.

Maybe it's time to combine efforts and merge repositories? 👀

**Comment by sumghai** [comment]
I personally can't see a merger being viable in the foreseeable future.

Even if senior management is changing, I'm pessimistic that CE:FT patches will be allowed into CE, and not simply being deleted for political reasons.

CE and CE:FT did briefly collaborate on the 1.6+ ammo/armor stat rebalance, and while we were admittedly quite wary and tepid towards the CE team's overtures, we were also frustrated that one or two members of the CE team only worked from their own private copies of the stat spreadsheets, instead of collaborating on either the public CE or CE:FT stat sheets. 

CE and CE:FT also have different patching philosophies:

* CE prefers standard calibers, whereas CE:FT is more willing to create new ammo defs for unusual or exotic weapons when required (for instance, CE's version of the Android Tiers mod simply patched the Mech Rifle to fire 40mm grenades, whereas CE:FT created a 40mm Bofors ammunition definition more accurate to the intention of the original AT mod)

* CE patches tend to be structured inside hierarchies of subdirectories, whereas CE prefers a flatter structure of simply the folder with the mod name and all the xml files on the same sublevel (to ease patch development)

Finally, CE has historically had a slower release cycle, due to their more extensive quality control process involving a curated selection of approved contributors; CE:FT attempts to maintain a faster release cycle, to keep up with changes in third-party mods.

**Comment by LockdownX7** [comment]
Hi, current CE maintainer here, figured I should pop by and clear up a few points:

> Even if senior management is changing, I'm pessimistic that CE:FT patches will be allowed into CE, and not simply being deleted for political reasons.

I know there's been hostility between some members of the CE and FT teams in the past, and I apologize on behalf of CE for any grievances that may have come from that. In my mind, there's no benefit in having people at each other's throats, least of all in an open-source project that has a lot to gain from open collaboration, so if it'd be possible to put those incidents behind us and establish a healthier base for cooperation, I think that would be a net positive for everyone.

The current policy for contributions in CE is as simple as it could be: pull requests are reviewed by a CE team member, followed by approval, or request for changes if any are needed. That's it. Patches from CE:FT, or indeed any source, would be welcome if submitted to the main CE repository.

> we were also frustrated that one or two members of the CE team only worked from their own private copies of the stat spreadsheets, instead of collaborating on either the public CE or CE:FT stat sheets.

I assume you're referring to the fire/venom arrows, that was a mistake on my part. Those stats will be reviewed and updated soon. The intention *is* to have a single public-facing stat sheet.

> CE prefers standard calibers, whereas CE:FT is more willing to create new ammo defs for unusual or exotic weapons when required (for instance, CE's version of the Android Tiers mod simply patched the Mech Rifle to fire 40mm grenades, whereas CE:FT created a 40mm Bofors ammunition definition more accurate to the intention of the original AT mod)

CE's stance on this is that an ammo type should be included in Defs/Ammo if it's a caliber that's reasonably expected to be used by more than one mod. This includes all real-world ammo types, as well as some fictional ones that are sufficiently generic. All other ammo defs should be added conditionally via XML patch, only if the mod that needs them is present in the load order. The rationale for this is as follows:

Take for example, the Mastiff and X5 ammo types that you added into CE:FT specifically for the Android Tiers patch. With the current implementation, those defs are loaded and processed by the ammo injector regardless of whether Android Tiers is actually in the player's mod list. If you do this for every mod, you'll eventually end up with hundreds of ammo types slowing down the startup process, unnecesarily, because they're only meant to be used by 1 mod. In this scenario, a better implementation would be to register those ammo defs at the same time you're patching the rest of Android Tiers, so they're never added if the FindMod operation for Android Tiers yields false.

> CE patches tend to be structured inside hierarchies of subdirectories, whereas CE prefers a flatter structure of simply the folder with the mod name and all the xml files on the same sublevel (to ease patch development)

There's no enforced standard regarding this, we leave it up to the patch author to organize their files as they see fit. In fact, a good chunk of patches currently in CE are single directory.

> Finally, CE has historically had a slower release cycle, due to their more extensive quality control process involving a curated selection of approved contributors; CE:FT attempts to maintain a faster release cycle, to keep up with changes in third-party mods.

This is something we intend to address. While feature updates will continue going through a QA process before eventually making it into a major release, compatibility patches are considered ready for release as soon as their review is approved, and we'll have a more frequent update schedule specifically for patch releases.


I believe that covers everything. If you have any more questions or concerns, let me know! I can also be reached for direct contact as LockdownX7 in Discord or Steam.

**Comment by AthenaSulisMinerva** [comment]
Hi @LockdownX7, thanks for dropping by.

We'd like to apologize too if we've caused any offence by running this fork - CE:FT was borne out of the frustrations of not only @sumghai, @Freedom187 and myself, but also various modders who had trouble dealing with former CE management in the past.

I've been fairly hands-off since helping to establish this fork, and there has been some very minor divergence in the codebase. I believe @sumghai would be able to brief you on the specific issues that need to be handled if and when the merger happens.

FYI - I'd like to hold this discussion publicly on GitHub, for a number of reasons:
* I personally dislike Discord and Steam; I've only communicated privately with @sumghai and @Freedom187 via Google Hangouts, but any formal tasks or decisions happen on GitHub
* Key details can be lost in Discord's scrollback; I think a lot of the bad blood between CE and FT arose from uncontrolled private conversations, a la "he said, she said" 
* AFAIK, @sumghai has a rather precarious situation with his internet - he is currently based in a location which makes it very difficult to reliably access Discord and Google Services, while GitHub is currently not blocked; even then, he was even summoned by authorities once after someone else made a joke on here about his current circumstances

We're finishing up on another FT patch release over the next few weeks, but I think the change in CE's management, along with my own long-term observations so far, means this is definitely a good time to start assessing the viability of merging the fork back to the original.

**Comment by sumghai** [comment]
If FT is to be folded back into the upstream CE, the following needs to be addressed:

## Mod Content

* **Compatibility Patches exclusive to FT**
  * e.g. Chicken Plucker mods
  * These should be fairly straightforward to integrate

* **Compatibility Patches with diverging implementation in CE and FT**
  * e.g. Android Tiers, New Ratkin Plus
  * I suggest deferring to the CE version for races, bodies, pawntypes and melee weapons, and the FT version for ranged weapons and "vehicles"

* **Ammo Type Factors exclusive to FT**
  * e.g. Railgun Sabot, Ballistic Knife, Plasma, Plasma Bomb, Non-Lethal Tranquilizer, possibly various Ratkin ballista variants
  * Will require some coordination

* **Real-world/Generic Ammo exclusive to FT**
  * e.g. 7.65x20mm Longue, 20x102mm NATO, 70mm APKWS guided rockets
  * Should also be straightforward

* **Real-world/Generic Ammo variants added by FT that extends standard CE ones**
  * e.g. Non-Lethal Tranquilizer versions of the 9x19mm Parabellum and 7.62x51mm NATO, originally made to support Chicken Plucker's Metal Gear Solid mod(s)
  * Generic enough to be included in the main CE ammo library

* **Mod-exclusive Ammo variants added by FT**
  * e.g. X-5 cartridge and Mastiff Shotgun shell (Android Tiers), Ratkin Ballista bolt variants (NewRatkinPlus), Xenn Cells (Xenn Race)
  * Should be registered as patches instead, as per @LockdownX7 
  * (ASIDE: I did initially try to recommended approach of patching exclusive ammo only when the associated mod was found; however, I ended up just adding them directly to the library after encountering a nasty but now long-forgotten error involving Ratkin crossbow projectiles)

* **[**

**Comment by sumghai** [comment]
@LockdownX7 @AthenaSulisMinerva

**Comment by sumghai** [comment]
(Sorry, accidentally closed the issue in the middle of my response)

@LockdownX7 @AthenaSulisMinerva 

If FT is to be folded back into the upstream CE, the following needs to be addressed:

## Mod Content

* **Compatibility Patches exclusive to FT**
  * e.g. Chicken Plucker mods
  * These should be fairly straightforward to integrate

* **Compatibility Patches with diverging implementation in CE and FT**
  * e.g. Android Tiers, New Ratkin Plus
  * I suggest deferring to the CE version for races, bodies, pawntypes and melee weapons, and the FT version for ranged weapons and "vehicles"

* **Ammo Type Factors exclusive to FT**
  * e.g. Railgun Sabot, Ballistic Knife, Plasma, Plasma Bomb, Non-Lethal Tranquilizer, possibly various Ratkin ballista variants
  * Will require some coordination

* **Real-world/Generic Ammo exclusive to FT**
  * e.g. 7.65x20mm Longue, 20x102mm NATO, 70mm APKWS guided rockets
  * Should also be straightforward

* **Real-world/Generic Ammo variants added by FT that extends standard CE ones**
  * e.g. Non-Lethal Tranquilizer versions of the 9x19mm Parabellum and 7.62x51mm NATO, originally made to support Chicken Plucker's Metal Gear Solid mod(s)
  * Generic enough to be included in the main CE ammo library

* **Mod-exclusive Ammo variants added by FT**
  * e.g. X-5 cartridge and Mastiff Shotgun shell (Android Tiers), Ratkin Ballista bolt variants (NewRatkinPlus), Xenn Cells (Xenn Race)
  * Should be registered as patches instead, as per @LockdownX7 
  * (ASIDE: I did initially try to recommended approach of patching exclusive ammo only when the associated mod was found; however, I ended up just adding them directly to the library after encountering a nasty but now long-forgotten error involving Ratkin crossbow projectiles)

* **Ammo/Projectile Artwork added by FT**
  * Will also require some reorganization of the current textures subfolders

* **FT Unique Feature: `Graphic_StackCountRanged`**
  * Introduced in FT-004 patch release
  * Variant of the vanilla `Graphic_StackCount`
  * Changes the graphic displayed for an item based on the current stack count, but with customizable upper limit for each subgraphic taken from their respective filenames
  * Stop-gap implementation for multi-shot power cell type-ammo
  * e.g. given a custom ammo item with textures `MyCustomAmmo_15`, `MyCustomAmmo_499`, `MyCustomAmmo_500`:

      If the stack size is... | ...then display:
      --- | :---:
      1~15|`MyCustomAmmo_15`
      16~499|`MyCustomAmmo_499`
      500 (or greater)|`MyCustomAmmo_500`

* **FT Unique Feature: `ammoConsumedPerShotCount`**
  * Planned for the upcoming FT-013 patch release
  * Optional xml tag
  * Allows ranged weapons to consume multiple rounds of ammunition per shot or trigger pull
  * Backwards compatible with all existing ranged weapons patches that do not use this feature
  * e.g. The "Super Shotgun" from the DOOM mod (double-barrel shotgun that fires both barrels simultaneously with one trigger pull), hypothetical Metal Storm-style multi-barrel weapons/turrets

**Comment by sumghai** [comment]
## Documentation

* **Consolidating Stat Balance Sheets**
  * [CE:FT Guns (Infantry and Turrets)](https://docs.google.com/spreadsheets/d/1Z9ZBnwRQWCKQm18O4I9ja4mFwLVChkNNfam2ZeQTTa0/edit?usp=sharing)
    * Entries are often duplicated between mods, for ease of reference
  * [CE:FT Ammunition and Projectiles](https://docs.google.com/spreadsheets/d/1-y3KCuTM_vJnxw-gQPl5KIiz7ySd7fw9P6Wx4zB5bZg/edit?usp=sharing)
    * Organized by type
    * New entries should be inserted alphabetically under the correct type/category, rather than appended as the last tow
  * [CE:FT Melee Tools (Races and Weapons)](https://docs.google.com/spreadsheets/d/1qwqFj6B7g-ft5Lqml6oEriuntlgr0mgH77-G1Iuh2_c/edit?usp=sharing)
    * Somewhat less complete/accurate due to lack of documentation
  * [CE:FT Animal and Race](https://drive.google.com/open?id=1aPTBeA6H3ANc8vH4FH7YbZ9i7lWj24gzh2nlvL3UTYY)

* **Patching Guides/Tutorials**
   * I would be interested in writing (and illustrating) comprehensive guides on patching for CE
   * Topics would range from file organization to xml tag explanations and example implementations for common weapon/armor/pawn types

## Administrivia
* GitHub Issue templates for Bug Reporting and Patch Request 
  * This would streamline issue reporting by encouraging users to provide all the necessary info to help developers diagnose issues

* Mod Preview Image and Steam Workshop aesthetic update
  * I recall that @AthenaSulisMinerva really hates Sombrero Man and the current CE font
  * We could also update the Steam Workshop description with (image) links that automatically open up the correct Bug Reporting or Patch Request template on GitHub, for user convenience

* List of Mods that CE bundles support for
  * Practically speaking, this should be a separate Markdown document to the README
  * Will help cut down on "is x compatible with CE?" questions
  * No longer a CE vs CE:FT comparison chart

**Comment by LockdownX7** [comment]
Hey again! I'm afraid I have bad news: I have recently stepped down as CE's maintainer (short venture, I know). The reasons that lead to this decision are complex, but in summary, it became clear to me that NoImageAvailable intends to maintain top-level control over the project, and we simply disagreed in key aspects regarding management. With that in mind, the only right move for me was to step down so that a more suitable candidate may be found for the role.

My apologies for this setback, hopefully CE's next maintainer can pick up this interaction where we left off. Really appreciate the work you guys do. Cheers!

**Comment by zozilin** [comment]
Oh wow, I had this idea of proposing merging CE and CE:FT for some time and it seems like it's already been discussed in depth. 

Hello, I am a stand-in admin on CE discord. Anyways, as Lockdown said, the position of maintainer is open again. 

I'll put it simply - seeing as Lockdown was the last guy with comprehensive CE and C#/Harmony knowledge, NiA being hands off the project (yeah, I know, conflicting messages here) - I think it's high time to merge projects and forget that a divide ever existed.

Details can be discussed in whatever form. Probably will come to patch quality control and code review as the most important points of conflict. 
If some of you wish to join us on Discord, please do. I think we have a lot to discuss.

**Comment by AthenaSulisMinerva** [comment]
If NoImageAvailable is still retaining top level control over CE in any form, then I'm afraid we're going to be much less enthusiastic about the proposal to merge.

**Comment by zozilin** [comment]
There is no top level control. As I said, he handed off the project.

**Comment by AthenaSulisMinerva** [comment]
![Capture-1](https://user-images.githubusercontent.com/49829079/72768527-a1df0380-3c32-11ea-8262-308dc3c7659a.PNG)

@LockdownX7 @zozilin, please clarify NiA's (non-)involvement, even if the role is advisory or emeritus.

I also wish to explicitly clarify that @sumghai is currently based in **_China_**, where Discord and Google services are **_banned_**. He only has sporadic access to both on certain days of the week for very short periods of time, and so he can't hold long conversations with me on Hangouts, much less with anyone else on Discord.

Github is the only reliable means of communication for the FT team, so I'm afraid designating a "Discord representative " as a point of contact will not be practical.

Therefore, we should continue to assess the viability of the merger publicly here on Github.

**Comment by zozilin** [comment]
Right, if Lockdown is up for talking about it, I'll speak of what happened yesterday too. 

But as I said, NiA is hands off the project. He still has owner rights on discord and repo, but has unsibscribed and muted all notifications. The only reason for him to return would be if someone went ballistic and started vandalizing a repo or something. And that if only someone contacted him about it. Which, coincidentally, happened yesterday. 

The maintainer is supposed to be the one calling the shots on project development. And now, as NiA and Lockdown left, there's no one to fill the role currently. Thus the proposal to merge and a request for someone with C# knowledge to be in that role.

Github is okay as well, whatever is needed to make this happen.

**Comment by sumghai** [comment]
I am not fit to be the maintainer or primary coder - I've only ever made minor C# modifications on an ad-hoc basis to quickly implement experimental features, and my unfamiliarity with much of the codebase was one of many points of contention between myself and the former CE leadership.

**Comment by AthenaSulisMinerva** [comment]
I'm even less knowledgeable about C# than sumghai - I've probably only recompiled the DLL a few times, and maybe changed one or two debug log messages.

@zozilin - We'll keep this open for the time being, but let us know when you've found a permanent maintainer, so that we have someone definitive to speak to.

**Comment by AthenaSulisMinerva** [comment]
@zozilin @LockdownX7 @N7Huntsman - have you guys finalized who will be assuming final leadership of the original CE mod?

**Comment by N7Huntsman** [comment]
Sorry about the delayed reply, we've sort of been discussing the matter amongst ourselves as well.

The short answer is no. The longer answer is that the combination of familiarity with CE's backend, technical skills, and interest in the project that we're looking for in a maintainer isn't something that's easy to scare up on demand. 

We've evaluated our options in regards to people who are currently with the project (such as Lockdown, who did a fine job despite his brief tenure), but the vast majority are nonstarters for one reason or another. Possible candidates from other circles aren't forthcoming, but we remain open to the possibility.

Ultimately, it looks like the project will be without a true maintainer for the immediate future, and no definitive timeline of when we expect to have one. 

While this naturally isn't ideal, it also isn't as damning as it may sound--the current development branch contains a number of important bug fixes and content updates, and we're generally of the mind that once it's pushed to master, Combat Extended will be on pretty solid footing. 

So, one prospect we're evaluating is making a concerted effort to get the fixes pushed to the main branch, and then being content to largely concerning ourselves with things like patches and balance.

From there, even with only a "lesser" maintainer (that is, someone who doesn't necessarily have the coding knowledge of a proper, permanent maintainer but who is capable of handing the Github aspects), the current contributors would be capable of continuing to add patches, maintain balance and spreadsheets, and handle various other .xml-centric concerns. While this would likely stifle the development of things like new mod mechanics or other coding tasks, it would offer us an opportunity to search for a contributor in a longer timeframe in an environment that's still productive in the meantime--including the prospect of merging FT patches and content.

**Comment by AthenaSulisMinerva** [comment]
Appreciate the update, @N7Huntsman.

For the next FT update, we'll start relocating mod-unique ammo types to patches, as well as consolidating damage/ammo types from both CE and FT, to help with the eventual transition to a merger.

**Comment by N7Huntsman** [comment]
That would be optimal. 

Ideally, when the merge comes around, the patches themselves will be essentially plug-and-play, and there won't too much tweaking that needs to be done in regards to balance or avoiding redundancy in so far as Defs/core files are concerned.

There's still a pending PR on Combat Extended with a bunch of ammo updates and the new types added to various calibers, but once that's accepted, you can safely assume all ammo and the ammo spreadsheet are 100%.

**Comment by sumghai** [comment]
@N7Huntsman, for mod-unique/exclusive ammo, would it make sense to also move the associated ammo item and projectile's custom graphics to, say, `Textures\Things\Ammo\**ThirdParty**` and `Textures\Things\Projectile\**ThirdParty**` respectively?

(Graphics for "shared" or "genericized" ammo and projectiles will remain in their appropriate type subfolders)

**Comment by N7Huntsman** [comment]
I might actually suggest structuring it like the patches folder: `Textures/ThirdParty, Core, ThirdParty, etc. ` That way, a patch for a mod consists of two folders, one under textures and one under Patches, and it's easier to add and remove stuff without having to dig through the rest of the content.

More broadly, in so far as graphics are concerned, it's important to consolidate textures where possible--whether that's sharing textures, or leaning on textures already packaged with the mod that's being patched (the latter generally being more for projectiles rather than ammo items). 

Since the game loads every single Def and Texture file on start-up regardless of whether they're being actively used by loaded content, with so many patched mods it's important to reduce loading time where possible. For Defs, stuff can be moved into a conditional patch, but there's not really much that can be done for textures aside from having fewer of them.

**Comment by sumghai** [comment]
> I might actually suggest structuring it like the Patches folder: `Textures/ThirdParty, Core, ThirdParty, etc. ` That way, a patch for a mod consists of two folders, one under textures and one under Patches, and it's easier to add and remove stuff without having to dig through the rest of the content.

So, something like this?:
* Example 1 - Android Tiers
  * `Patches/Android Tiers`
  * `Textures/ThirdParty/Android Tiers`
* Example 2 - [CP] DOOM and [CP] DOOM Kit Classic
  * `Patches/CP DOOM & DOOM Kit Classic`
  * `Textures/ThirdParty/CP DOOM & DOOM Kit Classic`

As for the genericized or "shared" CE ammo types, I suggest keeping their textures in their current locations, since moving them deeper into a `Core` subfolder would create far too much work for the CE team in changing path references for the vast majority of ammo.

> More broadly, in so far as graphics are concerned, it's important to consolidate textures where possible--whether that's sharing textures, or leaning on textures already packaged with the mod that's being patched (the latter generally being more for projectiles rather than ammo items).

Agreed.

The plasma power cells and railgun cartridges are good examples of formerly mod-unique ammo that were deemed generic enough to be added to the main CE:FT library.

> Since the game loads every single Def and Texture file on start-up regardless of whether they're being actively used by loaded content, with so many patched mods it's important to reduce loading time where possible. For Defs, stuff can be moved into a conditional patch, but there's not really much that can be done for textures aside from having fewer of them.

I see.

**Comment by N7Huntsman** [comment]
> So, something like this?
Sorry, to clarify I was just using "ThirdParty" as a placeholder name for any given mod, because that's what I thought you were doing. So it would be:

`Patches/Android Tiers`
`Textures/Android Tiers`

Additionally, I'll agree that it probably makes the most sense to leave core texture paths intact to avoid a bunch of needless .xml work.

**Comment by sumghai** [comment]
I suggested putting the third-party textures under a further `ThirdParty` subfolder, to keep the top-level `Textures` folder clean:

```
Textures/
├── Things/                 <-- core textures
│   ├── Ammo/
│   ├── Projectile/
│   └── (etc...)
├── ThirdParty
│   ├── Android Tiers/   <-- ammo (and projectiles, if not included in original mod) go here
│   ├── CP DOOM & DOOM Kit Classic/  <-- ditto
│   └── (etc...)
└── UI/                         <-- core UI textures
```

Otherwise, we'd have the following situation:

```
Textures/
├── Android Tiers/     <-- third party
├── CP DOOM & DOOM Kit Classic/    <-- third party
├── Things/                 <-- core textures
│   ├── Ammo/
│   ├── Projectile/
│   └── (etc...)
├── UI/                         <-- core UI textures
└── (other third-party texture folders)
```

**Comment by N7Huntsman** [comment]
As you like. My idea would've been fairly tidy with stuff tucked into a 'Core' folder, but I agree that it's simply not worth the tedious .xml change for minimal difference. My only real requirement is that it minimizes the number of locations content from a patch has to be distributed to--drop one folder in Patches and other somewhere in Textures, and you're good to go.

**Comment by sumghai** [comment]
Yup, sound good to me.

**Comment by N7Huntsman** [comment]
Additionally, as a general note for ammo, I think it will generally be alright (and indeed, preferable) to keep all real-world ammunition types in Core along with the ammo types deemed generic enough, rather than needing to place them within their patches. 

Obviously a lot of the mod-specific sci-fi and exotic ammo types are unique enough that we don't really have to worry about running into multiple patches using the same kind (unintentionally, at least), but there's always a chance when working with real life ammo, even with some of the more unusual rounds.

**Comment by cal066** [comment]
What is your policy on odd loads of real world ammo? eg HP .50 BMG, sabot 7.92 Kurz.
What are your thoughts on tandem rocket warheads? I am also trying to patch up Definitely More Cannons.

**Comment by NickZiggy** [comment]
Great, maybe feedback from players will actually count for something now. Because it certainly didn't since at least a year ago.

**Comment by N7Huntsman** [comment]
It is with no small amount of personal relief that I can report that all of the ammunition within Combat Extended is finally updated, which I can say with a fair amount of confidence was the last big item we were looking for have complete on our end before a merge to master and/or a release. With a little luck, playtesting on everything will go well, and we'll have something within a week or two.

**Comment by N7Huntsman** [comment]
@cal066 If you review the ammunition stat sheet, you'll notice that each weapon category has its own set of ammo types associated with it. Pistols get FMJ, HP, and AP; High Caliber get FMJ, AP-I, AP-HE, and Sabot; Rifles get the full gamut. So, in this case if you wanted to add HP to .50 BMG, you'd have to add it to all of the other High Caliber ammo sets as well, which isn't trivial.

The selection of ammo types isn't _entirely_ at random in this case; you weapons that are generally more expensive to manufacture (like dedicated sniper and anti-materiel rifles) and require more skilled shooters to get the most use out of predominately have access more capable, more expensive ammo types and the inverse is true for cheaper weapons (like pistols). Moreover, shooting something like Sabot or AP-HE out of some calibers just doesn't make sense because the benefit is so marginal, so those ammo types aren't available to those calibers.

So, ultimately, while I don't know that there's a formal "policy," but there are other considerations at work beyond making the most ammo available as possible.

Regarding tandem rocket warheads, I don't have strong feelings on them--they're not distinct in regards to how CE handles them compared to any other kind of high-AP rocket. I will remark though that most of the existing rockets of various types are generally more than capable of defeating the thickest armor they're likely to encounter--even the heaviest armor thicknesses are usually measured in 10s of millimeters, rocket penetration is measured in hundreds.

**Comment by AthenaSulisMinerva** [comment]
Thank you for the update, @N7Huntsman. I understand that @sumghai is still currently working to sync up the CE:FT stats sheets with the values from the CE sheet, as well as isolate the mod-exclusive ammo into patches.

**Comment by sumghai** [comment]
@N7Huntsman @zozilin @AthenaSulisMinerva - I have renamed this issue to better reflect the work CE:FT is undertaking in preparation for a (possible) merger back into CE.

I'm currently updating our own stat sheets with the new calculations from the CE version, as well as marking mod-unique ammo to be separated out later.

@N7Huntsman after CE:FT formally merges with CE, we may need to discuss how to handle exotic/sci-fi ammo/damage types (e.g. laser, plasma) - perhaps we could even create a separate subfeature/xml def type to handle power cell-like ammo items.

**Comment by N7Huntsman** [comment]
Would it be possible to pin this thread/issue somewhere? It's getting deeper in the Issue pile, but is probably an important enough piece of info for whoever else is tracking the merge/possibility thereof.

Currently, the CE Projectile sheet is 100% up-to-date, and I don't plan nor foresee needing to add any additional calculations or other stuff on our end in the immediate future, so it's probably safe for you to work off of it. At present, the only differences (in terms to calculations and the like) that I _notice_ between the sheets are the addition of Herbal Medicine as an ingredient for the poison arrows (which may also be useful for FT's tranq rounds), swapping the Stone blocks for Stone Chunks, and the addition of a Radiation ammo type, primarily for one particular advanced ammo.

In so far as I'm aware, the calculations for the other sheets like Melee, Animal Stats, and Guns haven't changed, though I know I've made a few tweaks to a few animal melee stats.

I agree that the exotic/sci-fi ammo types are likely to be the most intensive to get developed and balanced. I suppose we'll just sort of have to put our heads together and come up with something--in the meantime, we can probably leave whatever currently exists in place (presuming it's at least functional) and focus on integrating the patches that just sort of need a quick pass for balance (apparel, factions, real/historic weapons, etc.). With any luck, by the time it gets to working on the exotic weapons, we'll have come up with something.

Regarding the power cell subfeature you mentioned, I presume you're talking about something similar to what FT attempted with being able to dictate a custom value in the .xml for how much ammo is used per shot? It's an interesting idea and would certainly have quite a bit of utility.

**Comment by sumghai** [comment]
>Would it be possible to pin this thread/issue somewhere? It's getting deeper in the Issue pile, but is probably an important enough piece of info for whoever else is tracking the merge/possibility thereof.

Good idea! Just pinned it now.

> Currently, the CE Projectile sheet is 100% up-to-date, and I don't plan nor foresee needing to add any additional calculations or other stuff on our end in the immediate future, so it's probably safe for you to work off of it. At present, the only differences (in terms to calculations and the like) that I notice between the sheets are the addition of Herbal Medicine as an ingredient for the poison arrows (which may also be useful for FT's tranq rounds), swapping the Stone blocks for Stone Chunks, and the addition of a Radiation ammo type, primarily for one particular advanced ammo.

I'll update the CE:FT working copy with these bits now.

> In so far as I'm aware, the calculations for the other sheets like Melee, Animal Stats, and Guns haven't changed, though I know I've made a few tweaks to a few animal melee stats.

Okie doke

> I agree that the exotic/sci-fi ammo types are likely to be the most intensive to get developed and balanced. I suppose we'll just sort of have to put our heads together and come up with something--in the meantime, we can probably leave whatever currently exists in place (presuming it's at least functional) and focus on integrating the patches that just sort of need a quick pass for balance (apparel, factions, real/historic weapons, etc.). With any luck, by the time it gets to working on the exotic weapons, we'll have come up with something.

The current CE:FT implementations are more-or-less functional, so they can be brought over with minimal rebalance. As you said, we can worry about refining it later.

> Regarding the power cell subfeature you mentioned, I presume you're talking about something similar to what FT attempted with being able to dictate a custom value in the .xml for how much ammo is used per shot? It's an interesting idea and would certainly have quite a bit of utility.

The ammo-used-per-shot feature is mostly stable now, but yes, the power cell feature is tangentially related to it.

Once CE:FT is officially merged with CE, I will submit a proper feature request, but essentially I'd like to implement an xml framework to allow modders/patch authors to create any type of power cell ammo item:
* Compatible with the current AmmoUser system
* Each power cell item will have a comp counting the total shots left
* (Default) cells have fixed mass
* (Optional) cells my decrease in mass by a set amount after every shot (applicable to plasma or napalm tanks)
* For now, cells should be disposable (faction-specific charging stations are too much of a hassle)
* Projectile stats will continue to be calculated using the existing spreadsheet

**Comment by sumghai** [comment]
@N7Huntsman - I have updated the CE:FT ammo stat sheet as follows:

* Replaced stone blocks with stone chunks in crafting calculations
* Added Radiation ammo type factor
* Removed Ballistic Knife ammo type factor
* Reduced Non-Lethal (Tranq) damage to 10%, which comes out to 1 Tranquilizer damage for the tranq bullets
* Allow Non-Lethal (Tranq) ammo to be also crafted using Herbal Medicines; currently set to same ratio as Venom Arrows, which looks reasonably balanced given we usually craft 500+ rounds using 12~20 Herbs vs 10 arrows using 1~3 Herbs
* Moved some low-caliber Shells to High Caliber (<30 mm) category
* Identified the following ammo as being mod-unique:
  * **[Android Tiers]** Mastiff Shell
  * **[Android Tiers]** X-5 Cartridge
  * **[DOOM Classic]** UAC Energy Cell (Classic BFG 9000)
  * **[DOOM Classic]** UAC Energy Cell (Classic Plasma Gun)
  * **[DOOM Classic]** UAC Rocket Mk I
  * **[DOOM]** UAC BFG Cell (2016 BFG 9000)
  * **[DOOM]** UAC EMG Cell (2016 EMG Mk V Pistol)
  * **[DOOM]** UAC EMG Cell (2016 EMG Pistol)
  * **[DOOM]** UAC Plasma Cell (Gauss Rifle)
  * **[DOOM]** UAC Plasma Cell (Plasma Rifle/Gun)
  * **[DOOM]** UAC Plasma Cell (Vortex Rifle)
  * **[DOOM]** UAC Rocket Mk VI
  * **[Forsakens]** Forsakens Arrow
  * **[More Vanilla Turrets]** 81mm Mortar (Plasma Bomb)
  * **[NewRatkinPlus]** Ratkin Ballista Bolt
  * **[NewRatkinPlus]** Ratkin Ballista Bolt (AP)
  * **[NewRatkinPlus]** Ratkin Ballista Bolt (Heavy)
  * **[Pratt WWII]** Grenade F1
  * **[Pratt WWII]** Mills Grenade
  * **[Pratt WWII]** Mk II Grenade
  * **[Pratt WWII]** RGD-33
  * **[Pratt WWII]** Stielhandgranate 24
  * **[Pratt WWII]** Type 97 Grenade
  * **[Red Horse/Rimmu-Nation]** Ballistic Knife Blade
  * **[Red Horse/Rimmu-Nation]** C4
  * **[Red Horse/Rimmu-Nation]** Orbital EMP airstrike
  * **[Red Horse/Rimmu-Nation]** RGD-5
  * **[Xenn]** Xenn Fuel Cell (Cold)
  * **[Xenn]** Xenn Fuel Cell (Foam)
  * **[Xenn]** Xenn Fuel Cell (Hot)

 I have already relocated most of the textures for the mod-unique ammo/projectiles to the `Textures\ThirdParty\[mod name]` folders.

My next steps are:
* [ ] Rebalance all generic and mod-unique CE:FT ammo to the latest stat calculations 
* [ ] Re-implement mod-unique ammo as conditional PatchOps within each mod's compatibility patch

**Comment by AthenaSulisMinerva** [comment]
Good stuff @sumghai 

@N7Huntsman Would you recommend CE:FT point its upstream to CE's dev branch instead of master? That way, we can make sure we are synced up with the ammo (plus other) changes, and minimize conflicts when we merge back upstream to CE.

**Comment by N7Huntsman** [comment]
@sumghai Looks pretty good. As a note, the Radiation stats should probably be lumped in with the other sci-fi/exotic ammo as something that's a placeholder. I threw it together for one ammo in particular on the sheet, so it likely warrants another look.

Do all of those mod unique ammos have unique textures associated with them, or only some? Because as I'd mentioned, I'm keen to get as much mileage out of textures either already within CE or contained within the base mods that are being patched, and I'm thinking of a few on that list that can probably use generic/existing ammo textures.

@AthenaSulisMinerva I think the general procedure is that PRs that only affect Patches can go directly to the master branch, while changes to the "core" files (that is, any files not within the Patches folder) go to the development branch. So, the initial merge would go to development, and subsequent patches being added could go to master.

**Comment by sumghai** [comment]
> As a note, the Radiation stats should probably be lumped in with the other sci-fi/exotic ammo as something that's a placeholder. I threw it together for one ammo in particular on the sheet, so it likely warrants another look.

Noted. We can, of course, deal with sci-fi ammo in more deal after the first post-merger release of CE.

> Do all of those mod unique ammos have unique textures associated with them, or only some? Because as I'd mentioned, I'm keen to get as much mileage out of textures either already within CE or contained within the base mods that are being patched, and I'm thinking of a few on that list that can probably use generic/existing ammo textures.

Mod-unique ammo with unique textures hosted by CE itself:
  * **[Android Tiers]** Mastiff Shell
  * **[Android Tiers]** X-5 Cartridge
  * **[DOOM Classic]** UAC Energy Cell (Classic BFG 9000)
  * **[DOOM Classic]** UAC Energy Cell (Classic Plasma Gun)
  * **[DOOM Classic]** UAC Rocket Mk I
  * **[DOOM]** UAC BFG Cell (2016 BFG 9000)
  * **[DOOM]** UAC Plasma Cell (Gauss Rifle)
  * **[DOOM]** UAC Plasma Cell (Plasma Rifle/Gun)
  * **[DOOM]** UAC Plasma Cell (Vortex Rifle)
  * **[DOOM]** UAC Rocket Mk VI
  * **[More Vanilla Turrets]** 81mm Mortar (Plasma Bomb)

Mod-unique ammo that use generic CE ammo textures:
  * **[DOOM]** UAC EMG Cell (2016 EMG Mk V Pistol)
  * **[DOOM]** UAC EMG Cell (2016 EMG Pistol)
  * **[Xenn]** Xenn Fuel Cell (Cold)
  * **[Xenn]** Xenn Fuel Cell (Foam)
  * **[Xenn]** Xenn Fuel Cell (Hot)
  
Mod-unique ammo whose textures are provided externally by the original base mods themselves:
  * **[Forsakens]** Forsakens Arrow
  * **[NewRatkinPlus]** Ratkin Ballista Bolt
  * **[NewRatkinPlus]** Ratkin Ballista Bolt (AP)
  * **[NewRatkinPlus]** Ratkin Ballista Bolt (Heavy)
  * **[Pratt WWII]** Grenade F1
  * **[Pratt WWII]** Mills Grenade
  * **[Pratt WWII]** Mk II Grenade
  * **[Pratt WWII]** RGD-33
  * **[Pratt WWII]** Stielhandgranate 24
  * **[Pratt WWII]** Type 97 Grenade
  * **[Red Horse/Rimmu-Nation]** Ballistic Knife Blade
  * **[Red Horse/Rimmu-Nation]** C4
  * **[Red Horse/Rimmu-Nation]** Orbital EMP airstrike
  * **[Red Horse/Rimmu-Nation]** RGD-5

Looking at the first list, we can already see that only a small subset of mod-unique ammo have unique textures that need to be hosted by CE itself. These are from base mods whose authors are either unable or unwilling to host any CE-related content themselves, and the unique textures were created because otherwise using generic CE textures would detract from those specific base mods' intended player experience (especially the DOOM ammo, whose item textures are specific recreations of the iconic in-game items).

But even within the first list, there is significant texture sharing, mainly because they are simply projectile variants for a smaller set of actual ammo types - for example, the UAC Plasma Cell from the DOOM 2016 mod is actually used by four different weapons, each with four different projectile effects (with projectile graphics provided by the base mod). So really, the first list is already cut down to:
  * **[Android Tiers]** Mastiff Shell
  * **[Android Tiers]** X-5 Cartridge
  * **[DOOM Classic]** UAC Energy Cell (Classic BFG 9000 / Classic Plasma Gun projectile variants)
  * **[DOOM Classic]** UAC Rocket Mk I
  * **[DOOM]** UAC BFG Cell (2016 BFG 9000)
  * **[DOOM]** UAC Plasma Cell (Gauss Rifle, Plasma Rifle, Plasma Gun, Vortex Rifle projectile variants)
  * **[DOOM]** UAC Rocket Mk VI
  * **[More Vanilla Turrets]** 81mm Mortar (Plasma Bomb)

(ASIDE: I have also created other, additional ammo textures for CE:FT, but those are intended for generic use)

**Comment by AthenaSulisMinerva** [comment]
@N7Huntsman - Okay, so we'll temporarily switch our upstream to CE dev branch, pull over the recent dev changes, fix all the differences (@sumghai) and make a big FT merger PR back to the CE dev branch.

I'm planning to close off requests from the CE:FT issue tracker so that we can fully focus on the merger. If we do work on any last-minute compatibility patches, those will go straight to CE master.

**Comment by N7Huntsman** [comment]
@sumghai Okay, that list of ammos that need new textures on the CE-side isn't nearly as long as I feared. I'd be interested to hear what additional generic ammo textures you have; I'm thinking if nothing else, we'd have use for a generic railgun, plasma, and laser ammo.

@AthenaSulisMinerva Hmm, you know the situation on your side better than I do, so I'll leave the final say to you on how exactly to fold it into the CE, but I had some thought of bringing it in with a series of PRs rather than just one big one.

The first PR would be the core files that needed to go into Development (textures, new integrated ammo types, etc.) perhaps accompanied by whatever patches are as 100% balanced and polished as you want them to be. After that, patches with similar "levels" of confidence are PRed in batches.

My intention is to keep the Steam release as up to date as possible with the master branch itself--patches accepted into the master can probably expect to be added to the Steam workshop version within a day or so. With that in mind, there's still the intention on our end for everything that comes in to have at least a quick QA pass, since in the long run it saves us the time and effort of trying to interpret complaints in the Steam comments to identify any issues that arise (which a few inevitably will, as is the nature of this sort of thing).

My concern is that if you submit them all at once in one big pile, the PR will not only take longer to review, but may find itself waiting on the least "ready" patches--which sounds likely to be stuff related to sci-fi/exotic weapons, since we're still working out some of the details and even if we pushed it through, it'd only be placeholder stuff besides. Given that the sci-fi stuff will require additional textures and damage types that we may or may not have at present, it wouldn't surprise me in the slightest if it takes a couple of weeks after the initial merge for us to have a final version of it of them.

So, my thought is that if you divvy the patches up based upon the amount of attention you expect they'll need, not only will it get the polished patches integrated faster, it will also help ensure the stuff that needs more time (like the sci-fi/exotic content) isn't rushed for the sake of other, more polished content that they're packaged with.

Finally, on a lighter note, someone pointed out that since the impending push of the development branch to the master branch being slated as version 1.9, the merger will be Combat Extended 2.0, which I think is rather fitting.

**Comment by AthenaSulisMinerva** [comment]
> I'll leave the final say to you on how exactly to fold it into the CE, but I had some thought of bringing it in with a series of PRs rather than just one big one.
>
> The first PR would be the core files that needed to go into Development (textures, new integrated ammo types, etc.) perhaps accompanied by whatever patches are as 100% balanced and polished as you want them to be. After that, patches with similar "levels" of confidence are PRed in batches.

I definitely get where you're coming from with this, but I'm afraid splitting all the CE:FT commits up into separate PRs is nearly impossible for us to do.

From what I've read, this would involve creating multiple branches on our end from the point where CE:FT first diverged from CE, and for each branch manually pick the commits to go into them, before submitting the PRs. Various instructions on the internet even suggested splitting up each file into individual commits, which I can't say is particularly practical.

If there is an easier way to just pick the current state of the subset of files needed for the first main PR, I'm all ears.

**Comment by sumghai** [comment]
> Okay, that list of ammos that need new textures on the CE-side isn't nearly as long as I feared. I'd be interested to hear what additional generic ammo textures you have; I'm thinking if nothing else, we'd have use for a generic railgun, plasma, and laser ammo.

Let me get back to you on this.

**Comment by N7Huntsman** [comment]
> If there is an easier way to just pick the current state of the subset of files needed for the first main PR, I'm all ears.

Frankly, I'm not very good with the intricacies of Github either. But, a simple, if manual, way to do it would probably just be to make another Github account, use it to make another fork of Combat Extended, invite yourselves as collaborators on the new fork, and then you can freely add the subsets of files on that fork.

**Comment by sumghai** [comment]
@AthenaSulisMinerva - I'm sure we can figure something out.

@N7Huntsman - I've just done some housekeeping by relocating some of the existing ammo to match the upstream CE repo, rebalanced various crafting stats and segregated textures for mod-unique ammo.

Re-implementing the mod-unique ammo defs as patches themselves will take a while, as those will involve some playtesting to make sure they work. (I once encountered a nasty string of errors when I tried conditional patching for Ratkins Ballista ammo in the past, hence why I originally just threw everything into the CE library).

In the meantime, here's a list of genericized/real-world ammo and projectile textures added by CE:FT:
* 120 mm Tank Cannon shells (HE/HEAT/Sabot/Canister)
  * Generic enough to also be used for late-Cold War and Modern 105 and 125 mm tank shells
  * Sabot and Canister types currently not yet implemented
  * Canister shells are basically artillery shotgun shells, used by tanks to "scratch off" enemy infantry crawling over a friendly tank
* 762x385 mmR Cannon shells (HE/EMP)
* 50 mm GS-50 Grenades (HE/EMP)
  * Distinctive design due to the all-plastic construction of the rounds
* Bofors (AP/HE)
  * Shared by 40 and 57 mm Bofors shells
* Tomahawk Missile
* Non-lethal (Tranq) Pistol
* Non-lethal (Tranq) Rifle
* Plasma Heavy Power Cell
* Plasma Pistol Power Cell
* Plasma Rifle Power Cell
* Railgun High Caliber 
* Railgun Pistol
* Railgun Rifle
* 130 mm Type 63 Rocket (HE)
* 70 mm APKWS Rocket (HEAT) - essentially laser-guided Hydra 70 rockets

**Comment by N7Huntsman** [comment]
I glanced through the textures--everything looks fine, though I might suggest swapping the Bofors over to a shell rather than a high caliber if it hasn't been switched.

Additionally, is there anything on the C# side that you guys have add or changed as compared to base CE? I was asked to inquire.

**Comment by N7Huntsman** [comment]
@sumghai @AthenaSulisMinerva 

So you are aware, Combat Extended 1.9 has been released and is live on Steam and Github.

**Comment by AthenaSulisMinerva** [comment]
@N7Huntsman - I got a notification this morning.

@sumghai opened a PR to pull upstream commits into CE:FT, so that he can figure out where the merge conflicts are. I'll also leave it up to him how to merge our Core changes and Patches back upstream.

In the meantime, we're holding off on further FT releases, as a show of good faith and to focus on the merger.

**Comment by sumghai** [comment]
@N7Huntsman -  

> I glanced through the textures--everything looks fine, though I might suggest swapping the Bofors over to a shell rather than a high caliber if it hasn't been switched.

As in changing `<parent>AmmoHighCaliber</parent>` to `<parent>AmmoShells</parent>` in the xml defs for the two Bofors calibers? If so, that makes sense - I can do that tomorrow after work.

> Additionally, is there anything on the C# side that you guys have add or changed as compared to base CE? I was asked to inquire.

Off the top of my head:
* A new `Graphic_StackCountRanged` graphicClass that gives modders better control over what graphic is displayed for an item based on the current stack count; this works by retrieving the upper limit for each stack size interval from the filenames of the graphics themselves

* An optional `ammoConsumedPerShotCount` xml tag that allows ranged weapons to consume multiple rounds of ammunition per shot or trigger pull, such as double-barrelled shotguns or Metal Storm-style multi-barrel weapons turrets; setting this value to 0,1 or omitting it altogether will cause the weapon to default to standard CE behaviour of one ammo consumed per shot (ensuring backwards compatibility with all existing weapons patches

* Renaming of CE to CE:FT in debug log messages (this should be reverted as part of the merger)

@AthenaSulisMinerva - I'm going to try something with the `upstream_merger_core` branch.

**Comment by N7Huntsman** [comment]
I've asked one of the CE contributors familiar with C# to glance through FT's Assembly stuff in regards to merging everything, so I'll pass along anything of note he comes up with. You may see some minor bugfixes and the like, but other than that, we're about as ready for the merger as we can be on our end. With that mind, it's up to you when you feel you're end is ready and want to pull the trigger--though I might recommend doing it on a Friday or Saturday to give people some extra free time to dig into it.

In the meantime, have you considered at all what the Fast Track team wants to do after the merger? Obviously FT won't really exist in the same way it currently does, but there are only so many people out there making CE compatibility patches, and I'd hate to see that go to waste just because of the merger.

**Comment by AthenaSulisMinerva** [comment]
Personally, I'm going to take a step back from modding, but occasionally contribute patches if there's an obscure mod I really like.

@Freedom187 - We haven't heard from you in a while. What are your plans after FT gets merged into the upstream CE mod?

**Comment by sumghai** [comment]
@N7Huntsman - I had just about resolved all the merge conflicts on my end, and was almost ready to submit a PR for FT Core changes to the CE Dev branch last night.

However, at literally the last minute, some more commits popped up in CE Dev, resulting in even more merge conflicts that I have to now go back and manually fix.

I suspect it's mainly due to the different and inconsistent indentation in most of the affected files (CE:FT standardized on 4 spaces per indent for readability).

As for what happens after the merger, I'd be keen to continue maintaining patches, as well as work on the power cell feature and better wiki documentation/tutorials.

Also, for the hypothetical 2.0 update, I'd like to propose updating the mod preview image, adding subheading graphics to the Steam Workshop description etc - I know for a fact that @AthenaSulisMinerva loathes sombrero hat guy with a passion, and I think that taking CE:FT's current promo artwork* (but removing all references to FastTrack Edition) would make for a nice crisp and modern visual update.

**Comment by zozilin** [comment]
@sumghai I am sorry about the additional work, that was a fix that in my mind shouldn't wait. I will restrain from further dev merges.
As for the image - not opposed to that, but sombrero is a tradition of sorts. Maybe add one that to FT image?

**Comment by N7Huntsman** [comment]
@sumghai As it happens, we had one of the older contributors who knows C# pop back up and the discord and he's been working on stuff that otherwise has been kind of lingering. So on the one hand, it's unfortunate that the mergers are complicating things for you, but on the other it's nice these things are getting knocked out in pretty short order.

In regards to artwork, I'm certainly not opposed to a new look, but I'm inclined to agree with Zozillin that the sombrero is sort of traditional of CE--as absurd as the actual art is. Perhaps it stays or is redone in a similar style as a sort of "branding" sticker or something? Honestly, given that CE will need some lengthy work when Rimworld 1.1 releases to stable in a couple of weeks anyway, perhaps it's a matter we can push off for a bit.

**Comment by AthenaSulisMinerva** [comment]
@N7Huntsman @zozilin - I'm not really one for tradition, and @sumghai is being too diplomatic about it. I'm just going to say that yes, sombrero guy is too absurd, and it is an unpleasant reminder of various circumstances that lead to the creation of the FT fork in the first place.

But yes, we can discuss this as part of the big, extended push towards a 1.1-compatible update.

**Comment by sumghai** [comment]
I've just submitted CombatExtendedRWMod/CombatExtended/pull/1006, which only consists of Core changes and generic ammo additions - it seems that the merge issue has to do with the DLL, which necessitates detailed inspection of the source code to consolidate both sides' recent changes.

@N7Huntsman, @zozilin - feel free to take your time to go through the proposed changes and make sure they are up to snuff. I just wanted to make sure the Core FT stuff has been submitted rather than trapped in purgatory.

One last comment regarding promo art - the cowboy in the middle _could_ (no promises) be reworked into a version of sombrero guy. And while I have no intention of ever returning to the Discord channel, for the sake of consistency I could create a little white-on-black crossed revolver/musket logo to represent CE.

**Comment by sumghai** [comment]
Anyway, once 1006 gets approved, then comes the painstaking task of adding all the FT patches.

Some of them have diverging implementations from the upstream CE, in which case I will almost always defer to the original (for instance, the only thing I would change for Android Tiers would be to modify the weapons patches to instead use the conditionally-patched Mastiff and X-5 ammo).

Other patches, especially those involving animals, races and melee weapons, I will leave as-is, as those three were never really my forte.

**Comment by zozilin** [comment]
@sumghai, just a reminder, patch PRs usually go to master branch. Any ETA on the next step?

**Comment by sumghai** [comment]
@zozilin - My original intention, indeed, was to have PRs for compatibility patches go to CE's master branch, but some of them depend on generic ammo added as part of Phase I of the FT merger (which at the time of writing, are still only in CE's _development_ branch)

I can't give an ETA on my next steps, but what I'm definitely doing next is to manually go through all of CE:FT's patches (mod by mod) and merge them with CE's - generally, the CE implementation will take precedence, unless custom ammo from FT is involved.

A blue-sky request (which I may formally open a feature request issue ticket for) would be to implement power cell-type ammo items.

**Comment by Alicecomma** [comment]
Regarding the two C# additions by CE:FT
----

Adding a multi-barrel shotgun firing two shots at once can already be done as follows:
- VerbProperties.ticksBetweenBurstShots = -1
- VerbProperties.burstShotCount = 2
- (optional) CompProperties_FireModes.noSingleShot = true

While there is some talk of requiring multiple ammo resources per shot for futuristic cartridges, perhaps it is too speculative and it would be better to use the above, near-vanilla method and the near-CE method to allow for this behaviour. If futuristic cartridges are to be treated, perhaps much more involved methods are required, taking into account projectile damage, velocity, gun warmup and cooldown changes and the like.

----

Adding Graphic_StackCount with relatively specific images can also be done as follows, in Excel or Google Sheets:
- In cell A2 put a guessed subgraphics length
- In cell B2 put desired stackLimit
- In cell A4 put the value "1"
- In cell B4 put the formula "=IF(A4=1,0,IF(A4=$B$2,$A$2-1,round(min(1 + round(A4/$B$2*($A$2-3))+0.00001,$A$2-2))))"
- Drag cells A4, B4 down until stackLimit is reached in A504

Putting in a guessed subgraphics length, you can see whether the formula (used by Tynan as default in Graphic_StackCount.SubGraphicForStackCount(int stackCount, ThingDef def)) matches the desired index location. For example, 15-499-500 can be achieved with subgraphics length = 19 and stackLimit = 500. Sure, it would require 16 duplicate graphics, but it's possible to achieve.

I'm not exactly sure why such specific values must be reached by CE:FT. An alternative to the C# code would also be to simply PostFix SubGraphicForStackCount using Harmony patching. This way, existing behaviour of the method is guaranteed and additional behaviour is implemented without requiring CE-specific code.

**Comment by sumghai** [comment]
@Alicecomma 

> Adding a multi-barrel shotgun firing two shots at once can already be done as follows:

That seems to be quite convoluted, and less intuitive for modders.

I believe I attempted something similar originally, but as far as I remember it simply ended up firing two projectiles successively with a noticeable delay in-between, rather than a single shot that instantly consumes multiple units of ammo.

A number of other sci-fi energy weapons from the DOOM patches also already make use of `ammoConsumedPerShotCount`, so this FT feature isn't a one-off case made specifically for the double-barrelled shotgun.

> Adding Graphic_StackCount with relatively specific images can also be done as follows, in Excel or Google Sheets:

Using multiple, identical graphics with different filenames to shim up stack count graphic display is imprecise, added unnecessary assets during loading and inelegant.

**Comment by AthenaSulisMinerva** [comment]
I know very little about C#, @Alicecomma, but FT does have multiple, solid use cases for @sumghai's additions - they just happen to be in some of the patches yet to be ported over from FT to CE.

**Comment by N7Huntsman** [comment]
@Alicecomma I'm inclined to agree that that method seems rather hacky, and doesn't really achieve the ultimate objective of being able to have a single shot (regardless of how many projectiles are fired, one or multiple) that consumes a pre-determined amount of ammo. Having something along the lines of a `ammoConsumedPerShot` as described would be useful for something like a power cell--allowing for a single "power cell" ammo type to be used across a wider range of weapons.

**Comment by zozilin** [comment]
@sumghai Okay, my main concern was whether continuing other work on CE would interfere with your process right now and it doesn't seem to be the case.
I will resume reviewing and merging PRs not related to FT, but do tell me if that would create merging conflicts for you again.

**Comment by sumghai** [comment]
@zozilin - Yes, you guys should be fine resuming reviewing and merging non-FT PRs.

The only issue is on my end regarding submitting PRs for patches from FT, since some depend on the Phase I Core FT content approved in CombatExtendedRWMod/CombatExtended/pull/1006, which is currently only available on the dev branch

Should I wait for the CE master branch to be updated with dev stuff, before I start submitting compatibility patch PRs?

**Comment by N7Huntsman** [comment]
Patches are only allowed to be submitted into the Master branch because they're usually fairly quick to test and as a means to streamline getting them into a release. It's an allowance rather than a fixed policy, so go ahead and submit the patches to the Dev branch rather than the Master, since it makes far more sense to do so.

**Comment by sumghai** [comment]
Now that the Core FT changes have been merged into upstream, compatibility patches and mod-specific ammo are being reviewed via PRs on a case-by-case basis. This issue can be closed now.

@AthenaSulisMinerva, perhaps we should shutdown our issue tracker as well over the next few days.

**Comment by sumghai** [comment]
Okay @sumghai - I will close the tracker this weekend, to give everyone time to read the message.

Please also remove the issue reporting links on Steam.
